src/wallet/wallet.cpp

```diff
@@ -558,6 +558,109 @@ void CWallet::UpgradeDescriptorCache()
+    std::map<OutputType, int> tmpl = {{OutputType::LEGACY, 0}, {OutputType::P2SH_SEGWIT, 0}, {OutputType::BECH32, 0}};
+
+    for (const auto& spkm : GetAllScriptPubKeyMans()) {
+        const DescriptorScriptPubKeyMan* desc_spkm = dynamic_cast<DescriptorScriptPubKeyMan*>(spkm);
+        assert(desc_spkm);
+        LOCK(desc_spkm->cs_desc_man);
+        WalletDescriptor w_desc = desc_spkm->GetWalletDescriptor();
+
+        // Automatically generated descriptors are ranged
+        if (!w_desc.descriptor->IsRange()) {
+            continue;
+        }
+
+        // Automatically generated descriptors have exactly 1 xpub and no other keys
+        std::set<CPubKey> desc_pubkeys;
+        std::set<CExtPubKey> desc_xpubs;
+        w_desc.descriptor->GetPubKeys(desc_pubkeys, desc_xpubs);
+        if (desc_xpubs.size() != 1 || desc_pubkeys.size() != 0) {
+            continue;
+        }
+        const CExtPubKey& xpub = *desc_xpubs.begin();
+
+        if (key_counts.count(xpub) == 0) {
+            key_counts.emplace(xpub, std::make_pair(tmpl, 0));
+        }
+
+        // Automatically generated descriptors have an output type that is one of legacy, p2sh-segwit, or bech32
+        std::optional<OutputType> output_type = w_desc.descriptor->GetOutputType();
+        if (!output_type.has_value()) {
+            continue;
+        }
+        if (tmpl.count(output_type.value()) == 0) {
+            continue;
+        }
+        key_counts[xpub].first[output_type.value()]++;
+
+        if (w_desc.creation_time > key_counts[xpub].second) {
+            key_counts[xpub].second = w_desc.creation_time;
+        }
+    }
+
+    // Find candidate xpubs
+    // These are the ones that are used in 2 pkh(), 2 sh(wpkh()), and 2 wpkh() descriptors
+    // If the wallet's descriptors have been rotated before, then we want to get the most recent ones
+    // as those contain the master key that we currently want to use.
+    uint64_t best_time = 0;
+    std::optional<CExtPubKey> best_xpub;
+    std::map<CKeyID, CExtPubKey> xpubs;
+    for (const auto& [xpub, info] : key_counts) {
+        const auto& [dtypes, desc_time] = info;
+        if (std::all_of(dtypes.begin(), dtypes.end(), [](const auto& p) { return p.second == 2;})) {
+            xpubs.emplace(xpub.pubkey.GetID(), xpub);
+            if (desc_time > best_time) {
+                best_time = desc_time;
+                best_xpub.emplace(xpub);
+            }
+        }
+    }
+
+    if (best_xpub) {
+        std::optional<CKey> key;
+        auto key_it = desc_keys.find(best_xpub->pubkey);
+        if (key_it != desc_keys.end()) {
+            key = key_it->second;
+        }
+        std::vector<unsigned char> crypted_key;
+        auto ckey_it = desc_crypt_keys.find(best_xpub->pubkey);
+        if (ckey_it != desc_crypt_keys.end()) {
+            crypted_key = ckey_it->second;
+        }
+        SetActiveHDKey(*best_xpub, key, crypted_key);
```

Comment:
In ce5495a4b27f:

Need to check `SetActiveHDKey()` output. If it fails, the procedure should fail.

Reply:
I've added an `Assume` here. There isn't really a valid failure scenario here other than write failure.